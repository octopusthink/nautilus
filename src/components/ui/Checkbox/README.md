A Checkbox allows users to toggle a selection. It can be used to select one or more items from a list, or to provide users a "toggle" to change a boolean setting in an app.

```jsx
<Checkbox defaultChecked>Lions</Checkbox>
<Checkbox>Tigers</Checkbox>
<Checkbox>Bears</Checkbox>
<Checkbox defaultChecked>Oh my!</Checkbox>
```

## Usage

Use this component when users need to select a one or more items from a list of options in a form/input screen before submitting that info.

A checkbox can also be used as a "toggle"—clicking it will instantly toggle a setting between on/off.

## Appearance

Checkboxes traditionally appear with the checkbox first, then the text of the checkbox. Don't override this behaviour—if you want left-aligned text and right-align checkboxes, use a `Toggle` component instead.

Disabled components will appear using your theme's "disabled" colours:

```jsx
<Checkbox disabled>I am not active</Checkbox>
<Checkbox checked disabled>Neither am I, even though I'm checked</Checkbox>
```

## Interaction

A checkbox can be disabled, which will visually signify that it is not available to the user, and will also disable any attached event handlers. If you want to attach event handlers to a disabled `Checkbox`, you'll need to attach it to a parent container component.

```jsx
<Checkbox onClick={() => alert('I have been clicked.')}>I am active</Checkbox>
<Checkbox disabled onClick={() => alert('This code will not fire.')}>I am not active</Checkbox>
```

## Voice & Tone

TBD

## Accessibility

Checkboxes must always have a label. Whatever is supplied as the `children` prop of a `Checkbox` will be used as its `<label>`:

```jsx
<Checkbox>I am the label</Checkbox>
```
