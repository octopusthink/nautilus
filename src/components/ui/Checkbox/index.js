import React, { forwardRef, useMemo, useState } from 'react';
import { css } from '@emotion/core';
import shortid from 'shortid';
import PropTypes from 'prop-types';

import { interfaceUI, toUnits } from '../../../styles';
import { useTheme } from '../../../themes';

export const Checkbox = forwardRef((props, ref) => {
  const { children, id, noMargin, ...otherProps } = props;
  const [generatedId] = useState(shortid.generate());
  const inputId = useMemo(() => {
    return id || generatedId;
  }, [generatedId, id]);
  const theme = useTheme();

  return (
    <div
      css={css`
        position: relative;

        /* Set external margins */
        ${!noMargin &&
          css`
            margin-bottom: ${toUnits(theme.spacing.margin.xxSmall)};
          `}
      `}
    >
      <input
        id={inputId}
        type="checkbox"
        ref={ref}
        css={css`
          position: absolute;
          z-index: 1;
          width: 44px;
          height: 44px;
          margin: 0;
          opacity: 0;

          &:checked + label {
            color: ${theme.colors.text.dark};
          }

          /* Hide the checkmark by default. */
          & + label::after {
            content: none;
          }

          /*Unhide the checkmark on the checked state*/
          &:checked + label::after {
            content: '';
          }

          &:focus + label {
            outline: 0.2rem solid ${theme.colors.accent.primaryLight};
            outline-offset: 0.4rem;
            padding-right: 0.8rem;
          }
        `}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...otherProps}
      />
      <label
        htmlFor={inputId}
        css={css`
          ${interfaceUI.small(theme)};
          color: ${theme.colors.text.default};
          position: relative;
          padding-left: 3.2rem;

          &::before,
          &::after {
            position: absolute;
          }

          /*Outer-box*/
          &::before {
            border: 2px solid ${theme.colors.neutral.black};
            background: ${theme.colors.neutral.white};
            content: '';
            display: inline-block;
            height: 2rem;
            width: 2rem;
            top: -1px;
            left: 0;
          }

          /*Checkmark*/
          &::after {
            left: 4px;
            top: 5px;
            content: '';
            display: inline-block;
            height: 6px;
            width: 14px;
            border-left: 3px solid ${theme.colors.accent.secondary};
            border-bottom: 3px solid ${theme.colors.accent.secondary};
            transform: rotate(-50deg);
          }
        `}
      >
        {children}
      </label>
    </div>
  );
});

Checkbox.defaultProps = {
  children: undefined,
  disabled: false,
  error: undefined,
  hint: undefined,
  id: undefined,
  labelId: undefined,
  noMargin: false,
  onBlur: undefined,
  onChange: undefined,
  onFocus: undefined,
  unstyled: false,
};

Checkbox.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  onBlur: PropTypes.func,
  /** @ignore */
  onFocus: PropTypes.func,
  /** Disables this input; this applies a disabled style and disables user input/interaction with this element. This is useful if you have inputs that are conditionally allowed based on other states in your UI. */
  disabled: PropTypes.bool,
  /** An error message (either a simple string or a component) used to output an error message related to this component's value. If provided, an `aria-errormessage` will be set on the input component that will tell users of assistive technology the error message relates to this input. */
  error: PropTypes.node,
  /** HTML `id` attribute of the `input` element. Used for both the input's `id` attribute and the `<label>` `for` attribute. */
  id: PropTypes.string,
  /** Additional context to help users understand the purpose of the input. */
  hint: PropTypes.node,
  /** Visible text that serves to introduce the input. */
  label: PropTypes.node.isRequired,
  /** HTML `id` attribute for the `<label>` tag used to label the text input component. */
  labelId: PropTypes.string,
  /** Remove any outer margins from component. */
  noMargin: PropTypes.bool,
  /** `onChange` handler for the `TextField` component. */
  onChange: PropTypes.func,
  /* @ignore Don't output any CSS styles. */
  unstyled: PropTypes.bool,
};

export const { defaultProps, propTypes } = Checkbox;

export default Checkbox;
