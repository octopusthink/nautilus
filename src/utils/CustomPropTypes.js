import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
