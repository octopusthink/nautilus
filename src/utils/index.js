export { default as CustomPropTypes } from './CustomPropTypes';
