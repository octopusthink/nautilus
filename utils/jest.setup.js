// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
