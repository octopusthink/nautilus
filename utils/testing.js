// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
