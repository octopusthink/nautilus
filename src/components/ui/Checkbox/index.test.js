import React from 'react';

import { axe, fireEvent, render } from '../../../../utils/testing';
import Checkbox from '.';

describe('Checkbox', () => {
  it('should be wrapped in a <div>', () => {
    const { container } = render(<Checkbox>My checkbox</Checkbox>);

    expect(container.firstChild.tagName).toEqual('DIV');
  });

  it('should render an <input type="checkbox"> by default', () => {
    const { container } = render(<Checkbox>My checkbox</Checkbox>);

    expect(container.firstChild.firstChild.tagName).toEqual('INPUT');
    expect(container.firstChild.firstChild.getAttribute('type')).toEqual('checkbox');
  });

  it('should not be checked by default', () => {
    const { container } = render(<Checkbox>My checkbox</Checkbox>);

    expect(container.firstChild.firstChild.checked).toBe(false);
  });

  it('should respect the "defaultChecked" prop', () => {
    const { container } = render(<Checkbox defaultChecked>My checkbox</Checkbox>);

    expect(container.firstChild.firstChild.checked).toBe(true);

    fireEvent.click(container.firstChild.firstChild);

    expect(container.firstChild.firstChild.checked).toBe(false);
  });

  it('should match styles', () => {
    let { container } = render(<Checkbox id="my-checkbox">My checkbox</Checkbox>);

    expect(container.firstChild).toMatchSnapshot();

    // The label styles are altered when the checkbox is activated.
    container = render(
      <Checkbox id="my-checkbox" defaultChecked>
        My checkbox
      </Checkbox>,
    ).container;

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render a disabled <input type="checkbox"> when `disabled` prop is false', () => {
    const { container } = render(<Checkbox disabled={false}>My checkbox</Checkbox>);

    expect(container.firstChild.firstChild.getAttribute('disabled')).toEqual(null);
  });

  it('should render a disabled <input type="checkbox"> when `disabled` prop is set', () => {
    const { container } = render(<Checkbox disabled>My checkbox</Checkbox>);

    expect(container.firstChild.firstChild.getAttribute('disabled')).toEqual('');
  });

  it('should not output styles when the `unstyled` prop is set', () => {
    const { container } = render(
      <Checkbox id="unstyled-checkbox" unstyled>
        My checkbox
      </Checkbox>,
    );

    expect(container).toMatchSnapshot();
  });

  it("shouldn't have a margin when `noMargin` is true", () => {
    const { container } = render(
      <Checkbox id="no-margin-checkbox" noMargin>
        My checkbox
      </Checkbox>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <Checkbox className="custom-class" data-testid="myCheckbox">
        My checkbox
      </Checkbox>,
    );

    expect(getByTestId('myCheckbox').classList).toContain('custom-class');
  });

  it('should allow variables to manage the checked state', async () => {
    const { getByTestId } = render(
      <Checkbox data-testid="myCheckbox" checked onChange={() => {}}>
        My checkbox
      </Checkbox>,
    );

    expect(getByTestId('myCheckbox').checked).toBe(true);

    // Clicking this checkbox should not uncheck it, because its state
    // is managed by a variable.
    fireEvent.click(getByTestId('myCheckbox'));

    expect(getByTestId('myCheckbox').checked).toBe(true);
  });

  describe('accessibility', () => {
    it('should pass aXe tests', async () => {
      const { container } = render(<Checkbox>My checkbox</Checkbox>);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });

    it('should be toggled when clicked', async () => {
      const { container, getByTestId } = render(
        <Checkbox data-testid="myCheckbox">My checkbox</Checkbox>,
      );

      fireEvent.click(getByTestId('myCheckbox'));

      expect(container.firstChild.firstChild.checked).toBe(true);
    });
  });
});
