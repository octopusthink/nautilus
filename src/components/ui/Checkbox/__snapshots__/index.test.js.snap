// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should match styles 1`] = `
.emotion-0 {
  position: relative;
  margin-bottom: 0.4rem;
}

.emotion-1 {
  margin: 0;
  opacity: 0;
  position: absolute;
  z-index: 1;
}

.emotion-1:checked+label {
  color: #181b1c;
}

.emotion-1+label::after {
  content: '';
  opacity: 0;
  -webkit-transform: rotate(-35deg);
  -moz-transform: rotate(-35deg);
  -ms-transform: rotate(-35deg);
  transform: rotate(-35deg);
}

.emotion-1:checked+label::after {
  opacity: 1;
  -webkit-transform: rotate(-50deg);
  -moz-transform: rotate(-50deg);
  -ms-transform: rotate(-50deg);
  transform: rotate(-50deg);
}

.emotion-1:focus+label {
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
  padding-right: 0.4rem;
  padding-bottom: 0.2rem;
}

.emotion-2 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.7rem;
  line-height: 1.411764705882353;
  color: #3b3f45;
  position: relative;
  padding-left: 3.2rem;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
}

.emotion-2::before,
.emotion-2::after {
  position: absolute;
}

.emotion-2::before {
  border: 2px solid #181b1c;
  background: #ffffff;
  content: '';
  display: inline-block;
  height: 2rem;
  width: 2rem;
  top: -2px;
  left: 0;
}

.emotion-2::after {
  left: 4px;
  top: 4px;
  content: '';
  display: inline-block;
  height: 6px;
  width: 14px;
  border-left: 3px solid #0097a2;
  border-bottom: 3px solid #0097a2;
  -webkit-transition: opacity 0.1s linear,-webkit-transform 0.15s ease-in-out;
  transition: opacity 0.1s linear,transform 0.15s ease-in-out;
}

<div
  class="emotion-0"
>
  <input
    class="emotion-1"
    id="my-checkbox"
    type="checkbox"
  />
  <label
    class="emotion-2"
    for="my-checkbox"
  >
    My checkbox
  </label>
</div>
`;

exports[`Checkbox should match styles 2`] = `
.emotion-0 {
  position: relative;
  margin-bottom: 0.4rem;
}

.emotion-1 {
  margin: 0;
  opacity: 0;
  position: absolute;
  z-index: 1;
}

.emotion-1:checked+label {
  color: #181b1c;
}

.emotion-1+label::after {
  content: '';
  opacity: 0;
  -webkit-transform: rotate(-35deg);
  -moz-transform: rotate(-35deg);
  -ms-transform: rotate(-35deg);
  transform: rotate(-35deg);
}

.emotion-1:checked+label::after {
  opacity: 1;
  -webkit-transform: rotate(-50deg);
  -moz-transform: rotate(-50deg);
  -ms-transform: rotate(-50deg);
  transform: rotate(-50deg);
}

.emotion-1:focus+label {
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
  padding-right: 0.4rem;
  padding-bottom: 0.2rem;
}

.emotion-2 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.7rem;
  line-height: 1.411764705882353;
  color: #3b3f45;
  position: relative;
  padding-left: 3.2rem;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
}

.emotion-2::before,
.emotion-2::after {
  position: absolute;
}

.emotion-2::before {
  border: 2px solid #181b1c;
  background: #ffffff;
  content: '';
  display: inline-block;
  height: 2rem;
  width: 2rem;
  top: -2px;
  left: 0;
}

.emotion-2::after {
  left: 4px;
  top: 4px;
  content: '';
  display: inline-block;
  height: 6px;
  width: 14px;
  border-left: 3px solid #0097a2;
  border-bottom: 3px solid #0097a2;
  -webkit-transition: opacity 0.1s linear,-webkit-transform 0.15s ease-in-out;
  transition: opacity 0.1s linear,transform 0.15s ease-in-out;
}

<div
  class="emotion-0"
>
  <input
    checked=""
    class="emotion-1"
    id="my-checkbox"
    type="checkbox"
  />
  <label
    class="emotion-2"
    for="my-checkbox"
  >
    My checkbox
  </label>
</div>
`;

exports[`Checkbox should not output styles when the \`unstyled\` prop is set 1`] = `
<div>
  <div
    class="emotion-0"
  >
    <input
      class="emotion-0"
      id="unstyled-checkbox"
      type="checkbox"
    />
    <label
      class="emotion-0"
      for="unstyled-checkbox"
    >
      My checkbox
    </label>
  </div>
</div>
`;

exports[`Checkbox shouldn't have a margin when \`noMargin\` is true 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-1 {
  margin: 0;
  opacity: 0;
  position: absolute;
  z-index: 1;
}

.emotion-1:checked+label {
  color: #181b1c;
}

.emotion-1+label::after {
  content: '';
  opacity: 0;
  -webkit-transform: rotate(-35deg);
  -moz-transform: rotate(-35deg);
  -ms-transform: rotate(-35deg);
  transform: rotate(-35deg);
}

.emotion-1:checked+label::after {
  opacity: 1;
  -webkit-transform: rotate(-50deg);
  -moz-transform: rotate(-50deg);
  -ms-transform: rotate(-50deg);
  transform: rotate(-50deg);
}

.emotion-1:focus+label {
  box-shadow: 0 0 0 0.4rem #f5bbda;
  outline: 0;
  padding-right: 0.4rem;
  padding-bottom: 0.2rem;
}

.emotion-2 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.7rem;
  line-height: 1.411764705882353;
  color: #3b3f45;
  position: relative;
  padding-left: 3.2rem;
  -webkit-transition: all 0.1s linear;
  transition: all 0.1s linear;
}

.emotion-2::before,
.emotion-2::after {
  position: absolute;
}

.emotion-2::before {
  border: 2px solid #181b1c;
  background: #ffffff;
  content: '';
  display: inline-block;
  height: 2rem;
  width: 2rem;
  top: -2px;
  left: 0;
}

.emotion-2::after {
  left: 4px;
  top: 4px;
  content: '';
  display: inline-block;
  height: 6px;
  width: 14px;
  border-left: 3px solid #0097a2;
  border-bottom: 3px solid #0097a2;
  -webkit-transition: opacity 0.1s linear,-webkit-transform 0.15s ease-in-out;
  transition: opacity 0.1s linear,transform 0.15s ease-in-out;
}

<div
  class="emotion-0"
>
  <input
    class="emotion-1"
    id="no-margin-checkbox"
    type="checkbox"
  />
  <label
    class="emotion-2"
    for="no-margin-checkbox"
  >
    My checkbox
  </label>
</div>
`;
